File: backend/app/schemas/physical_io.py
from datetime import datetime

from pydantic import Field

from app.schemas.base import ApiSchema


SignalValue = bool | int | float | str


class MachineInputState(ApiSchema):
    emergency_stop: bool
    inspection_trigger: bool
    door_closed: bool


class PhysicalInputState(ApiSchema):
    revision: int = Field(ge=0)
    updated_at: datetime
    machine: MachineInputState
    sensors: dict[str, SignalValue]


class PhysicalOutputState(ApiSchema):
    revision: int = Field(ge=0)
    updated_at: datetime
    signals: dict[str, SignalValue]


class PhysicalOutputUpdate(ApiSchema):
    signals: dict[str, SignalValue]

File: frontend/src/types/physical-io.ts
export interface PhysicalInputState {
  revision: number;
  updatedAt: string;
  machine: {
    emergencyStop: boolean;
    inspectionTrigger: boolean;
    doorClosed: boolean;
  };
  sensors: Record<string, boolean | number | string>;
}

export interface PhysicalOutputState {
  revision: number;
  updatedAt: string;
  signals: Record<string, boolean | number | string>;
}